test/zdtm/static/membarrier.c

```diff
@@ -0,0 +1,105 @@
+#include <linux/membarrier.h>
+#include <sys/syscall.h>
+#include "zdtmtst.h"
+
+const char *test_doc = "Test membarrier() migration";
+const char *test_author = "Michał Mirosław <emmir@google.com>";
+
+/*
+ * Define membarrier() CMDs to avoid depending on exact kernel header version.
+ * FIXME: use MEMBARRIER_CMD_GET_REGISTRATIONS if supported by kernel.
+ */
+#define MEMBARRIER_CMD_PRIVATE_EXPEDITED			(1 << 3)
+#define MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED		(1 << 4)
+#define MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE		(1 << 5)
+#define MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE	(1 << 6)
+#define MEMBARRIER_CMD_PRIVATE_EXPEDITED_RSEQ			(1 << 7)
+#define MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_RSEQ		(1 << 8)
+
+static int membarrier(int cmd, unsigned int flags, int cpu_id)
+{
+	return syscall(__NR_membarrier, cmd, flags, cpu_id);
+}
+
+static const struct {
+	const char *name_suffix;
+	int register_cmd;
+	int execute_cmd;
+} membarrier_cmds[] = {
+	{ "",           MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED, MEMBARRIER_CMD_PRIVATE_EXPEDITED },
+	{ "_SYNC_CORE", MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE, MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE },
+	{ "_RSEQ",      MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_RSEQ, MEMBARRIER_CMD_PRIVATE_EXPEDITED_RSEQ },
+};
+static const int n_membarrier_cmds = sizeof(membarrier_cmds) / sizeof(*membarrier_cmds);
+
+static int check_membarriers(int barriers_registered)
+{
+	int ret = 0;
+
+	for (int i = 0; i < n_membarrier_cmds; ++i) {
+		if (~barriers_registered & membarrier_cmds[i].execute_cmd)
+			continue;
+		if (membarrier(membarrier_cmds[i].execute_cmd, 0, 0) < 0) {
+			pr_perror("membarrier(PRIVATE_EXPEDITED%s)", membarrier_cmds[i].name_suffix);
+			ret = 1;
+		}
+	}
+
+	if (ret) {
+		errno = 0;
+		fail("membarrier() cmds failed");
+	}
+
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int barriers_supported, barriers_registered, ret;
+
+	test_init(argc, argv);
+
+	barriers_supported = membarrier(MEMBARRIER_CMD_QUERY, 0, 0);
+	if (barriers_supported < 0) {
+		fail("membarrier() not supported by running kernel");
+		return 1;
+	}
+
+	barriers_registered = ret = 0;
+	for (int i = 0; i < n_membarrier_cmds; ++i) {
+		if (~barriers_supported & membarrier_cmds[i].register_cmd)
+			continue;
+		if (membarrier(membarrier_cmds[i].register_cmd, 0, 0) < 0) {
+			pr_perror("membarrier(REGISTER_PRIVATE_EXPEDITED%s)", membarrier_cmds[i].name_suffix);
+			ret = 1;
```

Comment:
Why can't we just fail immediately? I  mean switch `pr_perror` to `fail` and `ret = 1` to `return 1`.

Reply:
We can, but this is the same difference C compilers do (don't stop on first error) vs Pascal did (bail out on first failure). I prefer the former way as it minimizes the round-trips when debugging. Is this a hard requirement to follow the fail-early pattern?

Reply:
If you do some, let's say "trick", it must be meaningful - this way we improve code readability. If you don't strictly need something or don't know if you need something it is always better to avoid doing it, so that later reading the code we will not be confused that this "trick" is a must.

If something works in both variants we prefer the simple variant here in CRIU, at least me =) Because code is created to be easy to read.

If in future there would be a reason to debug this test in a manner that we need to try another membarrier if first membarrier call failed, e.g. if it would give us some new information for debugging, we can always improve the test, for now I don't see a clear reason.

Reply:
I've reworked the test a bit to make it more readable. Since this is a few lines change to support showing all registration/check compared to failing on first one, I think it's worth to have the extra info from the start. I see this a bit like insurance: you pay a bit more now, to reduce the amount of work in an emergency (here: a bug somewhere, not necessarily CRIU) in the future.